Feature/login view: 회원가입, 로그인, 학교검색 서버 통신 처리

hamfan524 opened the pull request:
- 제목 : Feat(#32 ): 기회원가입, 로그인, 학교검색 서버 통신 처리

## PR 유형

- [x] 새로운 기능 추가
- [ ] 버그 수정
- [x] 사용자 UI 디자인 변경
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [ ] 코드 리팩토링
- [ ] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [x] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## 🔎 작업 내용

- PhoneNumberViewModel
    - 전화번호를 바디에 넣어 서버에  요청 -> 실기기로 인증코드 메시지 발송
    - 받은 인증번호를 입력 후 한번 더 서버에 요청 -> 중복확인, 인증코드 인증
    - 전화번호값을 SignUpViewModel로 넘겨줌
- SignUpViewModel
    - userId, userName, birth, schoolData 값 SignUpViewModel에서 처리
    - 학교 검색 셀에서 텍스트필드 값 입력 시 0.5초 후 서버에 쿼리에 넣어서 요청
    - 받아온 데이터 테이블 셀에 뿌려줌 : 학교이름만 보여주던 셀에서 이름+주소 보여주는 셀로 디자인 수정
    - 선택 시 뷰모델에 학교데이터 보관
  <br/>

## 이미지 첨부
### 전화번호, 인증코드 인증

![Simulator Screen Recording - iPhone 15 Pro - 2024-07-15 at 18 26 00](https://github.com/user-attachments/assets/af3b163d-88b9-48b7-9281-1ccb90938727)

### 학교 검색, 선택
![Simulator Screen Recording - iPhone 15 Pro - 2024-07-15 at 18 29 44](https://github.com/user-attachments/assets/0dd638cf-54ef-45c6-907d-5db6a927d1ed)

hamfan524 committed:
🏫 Feat: 학교 데이터 요청 로직 구현

Domain
- SchoolDataRequest
- SchoolDataResponse
- SearchUsecase

Data
- SchoolDataDTO
- SchoolDataResponseImp

NetworkService
- SchoolDataEndPoint

문자열 데이터를 넣어 요청하면
문자열이 포함된 학교의 이름과 주소를 데이터로 받앙로는 로직 구현

Co-Authored-By: Lee ChangJun <109324421+Phangg@users.noreply.github.com>

hamfan524 committed:
🏫 Feat: 전화번호 입력, 인증 뷰모델 / 회원가입 뷰모델 분리

hamfan524 committed:
🏫 Feat: 회원가입-전화번호 인증 구현 완료

- 전화번호로 인증코드 전송
- 인증번호 재전송
- 전화번호, 인증번호 인증 요청 완료 -> 서버측 응답 데이터 수정 필요

hamfan524 committed:
🏫 Feat: 회원가입 - SignViewModel 연결

아이디 정규식 영어소문자 또는 숫자로 변경
유저이름, 생년월일, 학교 컬렉션뷰셀에 뷰모델 연결

이슈사항
- 학교검색 뷰에서 뷰모델 옵셔널값 못 읽어오는 현상 발견

hamfan524 committed:
🏫 Feat: 회원가입 - 학교 검색 기능 구현

입력된 문자열을 쿼리로 Api요청, 받아온 데이터에서 학교 이름만 테이블 셀로 보여주는 기능 구현

- 추가 구현해야하는 사항
입력 후 0.5초정도 가만히 있어야 통신 요청하는 로직으로 수정되어야함.

- 이슈사항
현재 입력 시 바로 요청이 되는게 아닌, 입력 후 한번 더 움직이면 그 이전 텍스트로 요청되는 현상 발생.

hamfan524 committed:
🏫 Feat: 회원가입 - 학교 검색

입력이 멈춘 후 0.5초 후 데이터 요청하도록 수정

- 이슈사항
데이터를 받아오면 바로 테이블뷰 셀이 업데이트 되는게 아닌, 한 번 더 움직여야 보이는 현상 유지

hamfan524 committed:
🐞 Fix: 회원가입 - 학교 검색

데이터를 받아오면 바로 테이블뷰 셀이 업데이트 되는게 아닌, 한 번 더 움직여야 보이는 현상 해결

hamfan524 committed:
🏫 Feat: 회원가입

PhoneNumberCollecionViewCell / UserIdCollecionViewCell / UserNameCollecionViewCell
- 텍스트필드에 문자를 입력하면 옆에 지우기 버튼 활성화

hamfan524 committed:
🏫 Feat: 학교 데이터 불러오는 동안 Progress보여주는 기능 구현

텍스트 필드에서 학교데이터 요청 시 progress돌아가게 구현

+ 추가 사항
 화면 전체 프로그레스 돌리는 프로토콜 파일 추가

hamfan524 committed:
🐞 Fix: 회원가입 - 학교 검색

xmark를 눌러 텍스트필드를 빈값으로 만들었는데,
filteredSchools 배열과 테이블뷰 셀들이 남아있는 오류 수정

hamfan524 committed:
🐞 Fix: 전화번호 인증 알람에서 뷰이동 관련 오류 수정

인증번호가 틀리거나, 중복된 계정인데도 회원가입 페이지로 넘어가는 오류 수정

Co-Authored-By: Noh Joo Young <96951827+JooYoungNoh@users.noreply.github.com>

hamfan524 committed:
🏫 Feat: 회원가입 - 학교 검색하면 학교이름과 주소도 같이 보여주도록 변경

셀 선택시 뷰모델에 학교 이름이 아닌 학교 데이터를 저장하도록 수정

hamfan524 committed:
🐞 Fix: 회원가입 - 텍스트필드 clear하면 한번 더 요청하는 오류 수정

JooYoungNoh reviewed (approved):
확인이여

ha-nabi reviewed (approved):
:)

Phangg commented on Projects/Feature/LoginFeature/Sources/SignUp/Presentation/Cell/SearchSchoolCollectionViewCell.swift:
```diff
@@ -355,16 +427,22 @@ extension SearchSchoolCollectionViewCell: UITextFieldDelegate {
     
     func textFieldDidChangeSelection(_ textField: UITextField) {
```
textFieldDidChangeSelection 메서드 로직에서
`디바운스콜` 을 사용하고 있는데, 아직은 아니지만, 다른 검색에서도 사용될 것 같아서요!

Debouncer 라는 class 를 하나 만들어서 사용하는 것도 편하겠다는 생각이 들어요!
찾아본 예시링크에요! [Debouncer 예시](https://magomercy.com/swift/iOS-%EC%95%B1-%EC%84%B1%EB%8A%A5-%ED%96%A5%EC%83%81%EC%9D%84-%EC%9C%84%ED%95%9C-Swift-Debounce-%EA%B5%AC%ED%98%84-%EB%B0%A9%EB%B2%95-c77489a0)
- 여기 예시에서는 DispatchWorkItem 대신, Timer 를 사용하더라구요.
둘의 성능면에서는 큰 차이가 없는 것 같은데, 
DispatchWorkItem 은 즉시 실행될 필요가 없는 짧은 지연 작업에 유리하고,
Timer 는 주기적인 작업 스케줄링 / 정기적인 UI 업데이트나 상태 체크 등과 같은 경우에 더 많이 사용된다고 합니다.
동호님도 한번 찾아보시면 더 좋을 것 같아요!!

hamfan524 replied:
넵 Debouncer 클래스 Common 프로젝트에 추가하여 사용하는 로직으로 변경하여 커밋했습니다!

hamfan524 committed:
🩹 Refactor: Debouncer 클래스 추가

Debouncer 클래스 생성 - 재사용 가능
- textFieldDidChangeSelection 메서드 로직에 Debouncer 사용
- DispatchWorkItem 대신, Timer 를 사용

hamfan524 committed:
💄 Design: 학교 검색 셀 padding값 변경

피그마에 새로 올라온 높이에 맞게 수정
